quantiles/test/quantiles_sketch_test.cpp

```diff
@@ -0,0 +1,899 @@
+    REQUIRE_THROWS_AS(sketch1.get_min_value(), std::runtime_error);
+    REQUIRE_THROWS_AS(sketch1.get_max_value(), std::runtime_error);
+    REQUIRE(sketch1.get_serialized_size_bytes() == 8);
+
+    const int n = 10000;
+    for (int i = 0; i < n; i++) sketch1.update(std::to_string(i));
+
+    REQUIRE(sketch1.get_min_value() == std::string("0"));
+    REQUIRE(sketch1.get_max_value() == std::string("9999"));
+
+    auto bytes = sketch1.serialize();
+    REQUIRE(bytes.size() == sketch1.get_serialized_size_bytes());
+    auto sketch2 = quantiles_string_sketch::deserialize(bytes.data(), bytes.size(), serde<std::string>(), 0);
+    REQUIRE(bytes.size() == sketch2.get_serialized_size_bytes());
+    REQUIRE(sketch2.is_empty() == sketch1.is_empty());
+    REQUIRE(sketch2.is_estimation_mode() == sketch1.is_estimation_mode());
+    REQUIRE(sketch2.get_n() == sketch1.get_n());
+    REQUIRE(sketch2.get_num_retained() == sketch1.get_num_retained());
+    REQUIRE(sketch2.get_min_value() == sketch1.get_min_value());
+    REQUIRE(sketch2.get_max_value() == sketch1.get_max_value());
+    REQUIRE(sketch2.get_normalized_rank_error(false) == sketch1.get_normalized_rank_error(false));
+    REQUIRE(sketch2.get_normalized_rank_error(true) == sketch1.get_normalized_rank_error(true));
+    REQUIRE(sketch2.get_quantile(0.5) == sketch1.get_quantile(0.5));
+    REQUIRE(sketch2.get_rank(std::to_string(0)) == sketch1.get_rank(std::to_string(0)));
+    REQUIRE(sketch2.get_rank(std::to_string(n)) == sketch1.get_rank(std::to_string(n)));
+  }
+
+  SECTION("sketch of strings, single item, bytes") {
+    quantiles_string_sketch sketch1(64, 0);
+    sketch1.update("a");
+    auto bytes = sketch1.serialize();
+    REQUIRE(bytes.size() == sketch1.get_serialized_size_bytes());
+    auto sketch2 = quantiles_string_sketch::deserialize(bytes.data(), bytes.size(), serde<std::string>(), 0);
+    REQUIRE(bytes.size() == sketch2.get_serialized_size_bytes());
+  }
+
+  SECTION("copy") {
+    quantiles_sketch<int> sketch1;
+    const int n(1000);
+    for (int i = 0; i < n; i++) sketch1.update(i);
+
+    // copy constructor
+    quantiles_sketch<int> sketch2(sketch1);
+    for (int i = 0; i < n; i++) {
+      REQUIRE(sketch2.get_rank(i) == sketch1.get_rank(i));
+    }
+
+    // copy assignment
+    quantiles_sketch<int> sketch3;
+    sketch3 = sketch1;
+    for (int i = 0; i < n; i++) {
+      REQUIRE(sketch3.get_rank(i) == sketch1.get_rank(i));
+    }
+  }
+
+  SECTION("move") {
+    quantiles_sketch<int> sketch1;
+    const int n(100);
+    for (int i = 0; i < n; i++) sketch1.update(i);
+
+    // move constructor
+    quantiles_sketch<int> sketch2(std::move(sketch1));
+    for (int i = 0; i < n; i++) {
+      REQUIRE(sketch2.get_rank(i) == (double) i / n);
+    }
+
+    // move assignment
+    quantiles_sketch<int> sketch3;
+    sketch3 = std::move(sketch2);
+    for (int i = 0; i < n; i++) {
+      REQUIRE(sketch3.get_rank(i) == (double) i / n);
+    }
+  }
+/*
```

Comment:
not needed?

Reply:
stolen from kll.. i can compute based on java and test or remove?